### ISSUE TITLE
Issue with circular references in CustomObjectComparer / EnumerableComparer

### ISSUE THREAD

**Issue by idealist1508** [GregFinzer/Compare-Net-Objects#237]
I implemented a `MyCustomTypeComparer` for my busines class and found an issue with circular references.
`MyCustomTypeComparer` looks basically like `EnumerableComparer`.  So I desided to test `EnumerableComparer`.

```c#
        [Test]
        public void CircularReferencesEnumerable()
        {

            var o1 = new ClassWithOverriddenHashCode();
            o1.Name = "1";
            IEnumerable<ClassWithOverriddenHashCode> e1 =(new [] {o1}).Where(_ => true);
            o1.MyCircularReference = e1;

            var o2 = new ClassWithOverriddenHashCode();
            o2.Name = "1";
            IEnumerable<ClassWithOverriddenHashCode> e2 =(new [] {o2}).Where(_ => true);
            o2.MyCircularReference = e2;

            CompareLogic logic = new CompareLogic();

            ComparisonResult result = logic.Compare(e1, e2);
            Assert.IsTrue(result.AreEqual, result.DifferencesString);
        }

```
This test fails

Fix is simple in this case.
```diff
diff --git a/Compare-NET-Objects/TypeComparers/EnumerableComparer.cs b/Compare-NET-Objects/TypeComparers/EnumerableComparer.cs
index 46f104c..945b6e1 100644
--- a/Compare-NET-Objects/TypeComparers/EnumerableComparer.cs
+++ b/Compare-NET-Objects/TypeComparers/EnumerableComparer.cs
@@ -40,16 +40,29 @@ namespace KellermanSoftware.CompareNetObjects.TypeComparers
         /// </summary>
         public override void CompareType(CompareParms parms)
         {
-            Type t1 = parms.Object1.GetType();
-            Type t2 = parms.Object2.GetType();
+            var oldObject1 = parms.Object1;
+            var oldObject2 = parms.Object2;
+            try
+            {
+                parms.Result.AddParent(parms.Object1);
+                parms.Result.AddParent(parms.Object2);
+
+                Type t1 = parms.Object1.GetType();
+                Type t2 = parms.Object2.GetType();
 
-            var l1 = TypeHelper.IsEnumerable(t1) ? ConvertEnumerableToList(parms.Object1) : parms.Object1;
-            var l2 = TypeHelper.IsEnumerable(t2) ? ConvertEnumerableToList(parms.Object2) : parms.Object2;
+                var l1 = TypeHelper.IsEnumerable(t1) ? ConvertEnumerableToList(parms.Object1) : parms.Object1;
+                var l2 = TypeHelper.IsEnumerable(t2) ? ConvertEnumerableToList(parms.Object2) : parms.Object2;
 
-            parms.Object1 = l1;
-            parms.Object2 = l2;
+                parms.Object1 = l1;
+                parms.Object2 = l2;
 
-            _compareIList.CompareType(parms);
+                _compareIList.CompareType(parms);
+            }
+            finally
+            {
+                parms.Result.RemoveParent(oldObject1);
+                parms.Result.RemoveParent(oldObject2);
+            }
         }
 
         private object ConvertEnumerableToList(object source)
```

How can I do the same in `MyCustomTypeComparer`?
I don't have access to the functions `parms.Result.AddParent(...)` and `parms.Result.RemoveParent(...)` from there.
Is it possible to change the accessibility level of these functions to public?

**Comment by GregFinzer** [comment]
@idealist1508 Certainly that would be fine.  Please create a pull request.

**Comment by GregFinzer** [comment]
Released in Version 4.74
https://github.com/GregFinzer/Compare-Net-Objects/releases/tag/v4.74

https://www.nuget.org/packages/CompareNETObjects/4.74.0
